torch/csrc/inductor/aoti_torch/shim_common.cpp

```diff
@@ -1672,6 +1675,234 @@ AOTITorchError aoti_torch_call_dispatcher(
+  auto self_type = current_schema.arguments()[0].type();
+  auto a_type = current_schema.arguments()[1].type();
+
+  ivalue_stack.push_back(to_ivalue(self_type, extension_stack[0])); // self
+  ivalue_stack.push_back(
+      to_ivalue(a_type, extension_stack[1])); // a (from extension)
+
+  // To get V3 to behave like V2: b=2 (fills with 2
+  ivalue_stack.push_back(c10::IValue(2));
+
+  return ivalue_stack;
+}
+
+} // namespace
+
+// Function to register test schema adapters for _test_schema_upgrader
+// This demonstrates the adapter registration pattern (internal use only)
+static AOTITorchError _register_test_adapters() {
+  // Register V1 to V3 adapter
+  if (auto err = register_schema_adapter(
+          "aten::_test_schema_upgrader",
+          VERSION_V2, // applies to versions < V2
+          adapt_v1_to_v3)) {
+    return err;
+  }
+
+  if (auto err = register_schema_adapter(
+          "aten::_test_schema_upgrader",
+          VERSION_V3, // applies to versions < V3
+          adapt_v2_to_v3)) {
+    return err;
+  }
+
+  // No adapter needed for extension_build_version >= V3
+  return AOTI_TORCH_SUCCESS;
+}
+
+// Static initialization to automatically register test adapters
+static struct TestAdapterInitializer {
+  TestAdapterInitializer() {
+    // Register the test adapters when the library loads
+    _register_test_adapters();
+  }
+} test_adapter_initializer;
+
+AOTI_TORCH_EXPORT AOTITorchError aoti_torch_call_dispatcher_v2(
+    const char* opName,
+    const char* overloadName,
+    StableIValue* stack,
+    // version of stable headers used to build the extension: necessary for
+    // applying schema adapters
+    uint64_t extension_build_version) {
+  AOTI_TORCH_CONVERT_EXCEPTION_TO_ERROR_CODE({
+    const auto op =
+        c10::Dispatcher::singleton().findSchemaOrThrow(opName, overloadName);
+    const auto& schema = op.schema();
+    const auto num_returns = schema.returns().size();
+    const auto num_arguments = schema.arguments().size();
+
+    torch::jit::Stack ivalue_stack;
+    auto& registry = SchemaAdapterRegistry::instance();
+
+    // Check if we need an adapter for this operation
+    if (auto adapter = registry.get_adapter(opName, extension_build_version)) {
+      // Use adapter to create IValue stack
+      ivalue_stack = (*adapter)(schema, stack, extension_build_version);
+    } else {
+      // No adapter needed - implementation matches aoti_torch_call_dispatcher
+      ivalue_stack.reserve(std::max(num_arguments, num_returns));
+      for (const auto idx : c10::irange(num_arguments)) {
+        auto stable_ivalue = stack[idx];
+        auto arg_type = schema.arguments()[idx].type();
+        torch::jit::push(ivalue_stack, to_ivalue(arg_type, stable_ivalue));
+      }
+    }
+
+    op.callBoxed(ivalue_stack);
+
+    // there should then be num_returns IValues on the stack, which
+    // we will convert to StableIValue and repopulate user input stack
+    for (const auto idx : c10::irange(num_returns)) {
```

Comment:
I'd anticipate there is a future possibility we may need adapters for returns too. From what I understand, the approach should be the same as with the inputs and thus can be added later without changing any APIs?